pkgs/applications/graphics/tesseract/default.nix

```diff
@@ -1,67 +1,91 @@
-{ stdenv, fetchFromGitHub, autoreconfHook, pkgconfig
+{ stdenv, fetchurl, fetchFromGitHub, autoreconfHook, pkgconfig
 , leptonica, libpng, libtiff, icu, pango, opencl-headers
 # Supported list of languages or `null' for all available languages
 , enableLanguages ? null
 # if you want just a specific list of languages, optionally specify a hash
 # to make tessdata a fixed output derivation.
 , enableLanguagesHash ? (if enableLanguages == null # all languages
-                         then "1h48xfzabhn0ldbx5ib67cp9607pr0zpblsy8z6fs4knn0zznfnw"
+                         then "11bi1hj2ihqrgvi9cam8mi70p4spm3syljkpnbglf4s8jkpfn15a"
                          else null)
 }:
 
-let tessdata = stdenv.mkDerivation ({
-  name = "tessdata";
-  src = fetchFromGitHub {
-    owner = "tesseract-ocr";
-    repo = "tessdata";
-    rev = "3cf1e2df1fe1d1da29295c9ef0983796c7958b7d";
-    # when updating don't forget to update the default value fo enableLanguagesHash
-    sha256 = "1v4b63v5nzcxr2y3635r19l7lj5smjmc9vfk0wmxlryxncb4vpg7";
-  };
-  buildCommand = ''
-    cd $src;
-    for lang in ${if enableLanguages==null then "*.traineddata" else stdenv.lib.concatMapStringsSep " " (x: x+".traineddata") enableLanguages} ; do
-      install -Dt $out/share/tessdata $src/$lang ;
-    done;
-  '';
-  preferLocalBuild = true;
-  } // (stdenv.lib.optionalAttrs (enableLanguagesHash != null) {
-  # when a hash is given, we make this a fixed output derivation.
-  outputHashMode = "recursive";
-  outputHashAlgo = "sha256";
-  outputHash = enableLanguagesHash;
-  }));
-in
+let
+  languages = (import ./languages.nix { inherit stdenv fetchurl fetchFromGitHub; }).v3;
 
-stdenv.mkDerivation rec {
-  name = "tesseract-${version}";
-  version = "3.05.00";
+  tessdata = if enableLanguages == null then
+      languages.all
+    else
+      if builtins.all (lang: builtins.hasAttr lang languages) enableLanguages then
+        map (lang: languages.${lang}) enableLanguages
+      else
+        # Copy the selected languages from languages.all
+        stdenv.mkDerivation ({
+          name = "tessdata";
+          buildCommand = ''
+            mkdir $out
+            cd ${languages.all}
+            cp ${stdenv.lib.concatMapStringsSep " " (x: x + ".traineddata") enableLanguages} $out
```

Comment:
Is there anything large enough _not_ copied in the downloaded language packs? If no, you could just `ln -s`, I guess?

Reply:
Yes. If we just need a single exotic language, like `cym`, all other languages (1.2 GB) are needlessly kept in the store.
The old behaviour was also to copy the files.

Reply:
Oops, my bad, I thought each language is copied from the individual language pack.

Reply:
In case you missed it: I've added a line to my previous reply.

Reply:
I don't understand that `if` now. «If we mention all languages, use individual derivations instead of `all`»?

If just a list of derivations selected by language names would work without copying, that sounds better than that extra composition step.

Reply:
Currently we're not providing all tessdata languages as separate drvs, just a selection of the most popular ones. 
So for more obscure langs we have to tap into `languages.all`.
Here's a translation of the second `if`-statement:
```
If all languages are available as separate drvs:
    build a list of drvs
else:
    build a new tessdata drv by copying all requested languages from languages.all
```

Reply:
Note, again, that this is hugely simplified if switch to seperate drvs for all langs. :smiley_cat:

Reply:
Now I understand what «switching to individual derivations for all languages» actually mean… Yes, sounds like a good idea. I thought it was about less support for multi-language builds.

Reply:
Alright, sorry if I wasn't perfectly clear about that.
I've added a [commit](https://github.com/NixOS/nixpkgs/pull/52379/commits/51af0271b0f19cca9824ef64207cf8233c846971) that implements separate lang drvs.
The evaluation time of `nixpkgs.tesseract` (which includes all languages) is increased by 6.8% (361ms vs 338ms, via [hyperfine](https://github.com/sharkdp/hyperfine)) but I think the added flexibility and simplicity is worth the little extra eval time. What do you think?

Reply:
Maybe default to the old list of languages? Would it bring the time back down?

Reply:
With the current approach we're evaluating one `fetchurl` drv for each language (107 total) and pass their store paths to the builder.

If we bring back a single fixed output drv for all langs, then the performance will be the same as before.

Reply:
Hm. Now I wonder if it is better to have a medium-size list of languages by default, or just download the huge all-languages tarball…

Reply:
(Yes, this might show you that it is not you being dense, this is also me being unsure and therefore not always picking words well. I guess I will trust your choice in this question, given that building a local three-language version will be cheap)

Reply:
Ah, my doubts were confirmed: Caching the 'all languages' tessdata with a fixed output drv that links all lang drvs into a single folder is not possible. That's because nix always fully evaluates all derivation attributes, even for fixed output drvs. So all lang drvs that are passed to the builder will still be evaluated.

We have the following options now:
1. Accept the slightly worse perf.
2. Cache 'all languages' with a drv that fetches the whole tessdata repo, like before. But langs are stored twice this way. (Hydra doesn't seem to use `nix-store --optimise`).
3. As proposed by you, restrict default langs for `nixpkgs.tesseract` to a smaller subset. But this breaks backwards compability. Previously, the full tessdata has always been included .

Reply:
I'm actually tending towards 3. What do you think?

Reply:
There is also the option 2.1, which fetches the whole repository and sets empty `meta.hydraPlatforms` for the individual languages. Then only English is stored twice on Hydra (used for VM tests). Arguably, we could even avoid keeping any language except English on Hydra… (option 2.2)

Reply:
Please see **"Edit 2"** in the main PR description for a short summary of our current approach. I think we've hit the sweet spot.